src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -84,44 +84,66 @@ size_t MonitorList::max() const {
   return Atomic::load(&_max);
 }
 
-// Walk the in-use list and unlink (at most MonitorDeflationMax) deflated
-// ObjectMonitors. Returns the number of unlinked ObjectMonitors.
+// Walk the in-use list and unlink deflated ObjectMonitors.
+// Returns the number of unlinked ObjectMonitors.
 size_t MonitorList::unlink_deflated(Thread* current, LogStream* ls,
                                     elapsedTimer* timer_p,
+                                    size_t deflated_count,
                                     GrowableArray<ObjectMonitor*>* unlinked_list) {
   size_t unlinked_count = 0;
   ObjectMonitor* prev = nullptr;
-  ObjectMonitor* head = Atomic::load_acquire(&_head);
-  ObjectMonitor* m = head;
+  ObjectMonitor* m = Atomic::load_acquire(&_head);
+
   // The in-use list head can be null during the final audit.
   while (m != nullptr) {
     if (m->is_being_async_deflated()) {
-      // Find next live ObjectMonitor.
+      // Find next live ObjectMonitor. Batch up the unlinkable monitors, so we can
+      // modify the list once per batch. The batch starts at "m".
+      size_t unlink_batch = 0;
       ObjectMonitor* next = m;
+      // Look for at most MonitorUnlinkBatch monitors, or the number of
+      // deflated and not unlinked monitors, whatever comes first.
+      size_t unlink_limit = MIN2<size_t>(deflated_count - unlinked_count, MonitorUnlinkBatch);
       do {
         ObjectMonitor* next_next = next->next_om();
         unlinked_count++;
         unlinked_list->append(next);
         next = next_next;
-        if (unlinked_count >= (size_t)MonitorDeflationMax) {
-          // Reached the max so bail out on the gathering loop.
+        if (unlink_batch++ >= unlink_limit) {
+          // Reached the max batch, so bail out of the gathering loop.
+          unlink_batch = 0;
+          break;
+        }
+        if (prev == nullptr && Atomic::load(&_head) != m) {
+          // Current batch used to be at head, but it is not at head anymore.
+          // Bail out and figure out where we currently are. This avoids long
+          // walks searching for new prev during unlink under heavy list inserts.
           break;
         }
```

Comment:
Is there a reason to make this check and bail out while executing the `do ... while`
and gathering the first batch of deflated monitors? Why can't you finish gathering
your first batch and let the head change logic below just take care of it?

Reply:
My thinking was two-fold: a) with very large batch and heavy inserts, it might take a while for us to break out of gathering loop, which would force us to look through a longer list for batch head in the unlink loop -- and since that eats up TTSP, I wanted to make sure we detect this as soon as possible; b) since `prev != nullptr` after the first update, we do not incur lots of overheads with checking in the gathering loop.

But you are right that we might not need this for small batch sizes, which is what we use to keep TTSP in control. I can remove it, if you want.

Reply:
I still don't get it, but I don't want to hold this PR up any longer.